packages/langium/src/serializer/json-serializer.ts

```diff
@@ -113,7 +130,12 @@ export class DefaultJsonSerializer implements JsonSerializer {
                 }
             }
         };
-        internalRevive(object);
-        return object;
+        internalRevive(root);
```

Comment:
This helper function is quite hard to understand. Here's an approach for simplifying it (more like pseudo-code, needs to be adapted):
```typescript
function linkNode(node: AstNode, root: AstNode, container?: AstNode, containerProperty?: string, containerIndex?: number): void {
    node.$container = container;
    node.$containerProperty = containerProperty;
    node.$containerIndex = containerIndex;
    for (const [propertyName, item] of Object.entries(node)) {
        if (Array.isArray(item)) {
            for (let index = 0; index < item.length; index++) {
                const element = item[index];
                if (isReference(element)) {
                    element.ref = getAstNode(root, element.$ref);
                } else if (isAstNode(element)) {
                    linkNode(element, root, node, propertyName, index);
                }
            }
        } else if (isReference(item)) {
            item.ref = getAstNode(root, item.$ref);
        } else if (isAstNode(item)) {
            linkNode(item, root, node, propertyName);
        }
    }
}
```